File: packages-js/bpmn-flowaudit/src/collection/collection.ts
/**
 * Mutable diagram collection with integrity checks on every change (same
 * operations as `Diagrammsammlung` of `auditcore_bpmn`). Each diagram lives
 * in exactly one folder (`folderId`; `null` = top level) and carries any
 * number of tags.
 */

import type { ProcessModel } from '../model/processModel'
import {
  COLLECTION_SCHEMA,
  KEY_KINDS,
  emptyExcerpt,
  isValidId,
  type CollectionData,
  type DiagramEntry,
  type Folder,
  type KeyKind,
  type Tag,
} from './collectionData'
import { excerptFromModel } from './excerpt'

const MESSAGES: Record<string, string> = {
  invalidId: 'Ungültige {what}-ID „{id}“ (erlaubt: Buchstaben, Ziffern, _ . -).',
  exists: '{what} „{id}“ existiert bereits.',
  unknownFolder: 'Ordner „{id}“ ist unbekannt.',
  unknownDiagram: 'Diagramm „{id}“ ist unbekannt.',
  unknownTags: 'Unbekannte Tags: {ids}',
  cycle: 'Ein Ordner kann nicht in sich selbst oder einen Unterordner verschoben werden.',
  notEmpty: 'Ordner „{id}“ ist nicht leer.',
  order: 'Die Reihenfolge muss genau die Diagramme des Ordners enthalten.',
  schema: 'Sammlungsschema {schema} wird nicht unterstützt.',
  unknownKind: 'Unbekannte Schlüsselart „{kind}“.',
  approvedImmutable: 'Version {version} von „{id}“ ist bereits freigegeben; Änderungen erfordern eine neue Version.',
  noApproval: 'Für „{id}“ gibt es keinen freigegebenen Stand.',
}

export class CollectionError extends Error {
  constructor(
    readonly code: string,
    readonly params: Record<string, string> = {},
  ) {
    super((MESSAGES[code] ?? code).replace(/\{(\w+)\}/g, (_m, key: string) => params[key] ?? ''))
  }
}

function byOrder<T extends { order: number; name: string; id: string }>(a: T, b: T): number {
  return a.order - b.order || a.name.localeCompare(b.name, 'de') || a.id.localeCompare(b.id)
}

export interface FolderNode {
  folder: Folder | null
  subfolders: FolderNode[]
  diagrams: DiagramEntry[]
}

export class DiagramCollection {
  id: string
  name: string
  readonly folders = new Map<string, Folder>()
  readonly tags = new Map<string, Tag>()
  readonly diagrams = new Map<string, DiagramEntry>()

  constructor(id = 'sammlung', name = 'Diagrammsammlung') {
    this.id = id
    this.name = name
  }

  // -- folders and tags -----------------------------------------------------
  private checkId(id: string, what: string): void {
    if (!isValidId(id)) throw new CollectionError('invalidId', { id, what })
  }

  private checkFolder(folderId: string | null | undefined): void {
    if (folderId && !this.folders.has(folderId)) throw new CollectionError('unknownFolder', { id: folderId })
  }

  createFolder(id: string, name: string, parentId: string | null = null, description?: string): Folder {
    this.checkId(id, 'Ordner')
    if (this.folders.has(id)) throw new CollectionError('exists', { id, what: 'Ordner' })
    this.checkFolder(parentId)
    const order = 1 + Math.max(-1, ...this.subfolders(parentId).map((f) => f.order))
    const folder: Folder = { id, name, parentId, order, ...(description ? { description } : {}) }
    this.folders.set(id, folder)
    return folder
  }

  renameFolder(id: string, name: string): void {
    this.folder(id).name = name
  }

  private folder(id: string): Folder {
    const folder = this.folders.get(id)
    if (!folder) throw new CollectionError('unknownFolder', { id })
    return folder
  }

  ancestors(folderId: string | null | undefined): string[] {
    const chain: string[] = []
    for (let current = folderId ?? null; current; current = this.folder(current).parentId ?? null) {
      if (chain.includes(current)) throw new CollectionError('cycle', { id: current })
      chain.push(current)
    }
    return chain
  }

  moveFolder(folderId: string, parentId: string | null, position?: number): void {
    const folder = this.folder(folderId)
    this.checkFolder(parentId)
    if (parentId && this.ancestors(parentId).includes(folderId)) throw new CollectionError('cycle')
    folder.parentId = parentId
    const siblings = this.subfolders(parentId).filter((f) => f.id !== folderId)
    siblings.splice(Math.max(0, Math.min(position ?? siblings.length, siblings.length)), 0, folder)
    siblings.forEach((item, index) => (item.order = index))
  }

  removeFolder(folderId: string): void {
    const hasChildren = this.subfolders(folderId).length > 0 || this.inFolder(folderId).length > 0
    if (hasChildren) throw new CollectionError('notEmpty', { id: folderId })
    this.folders.delete(folderId)
  }

  createTag(id: string, name: string, color?: string): Tag {
    this.checkId(id, 'Tag')
    const tag: Tag = { id, name, ...(color ? { color } : {}) }
    this.tags.set(id, tag)
    return tag
  }

  // -- diagrams ---------------------------------------------------------------
  addDiagram(id: string, options: { name?: string; folderId?: string | null; tags?: string[]; model?: ProcessModel } = {}): DiagramEntry {
    this.checkId(id, 'Diagramm')
    if (this.diagrams.has(id)) throw new CollectionError('exists', { id, what: 'Diagramm' })
    this.checkFolder(options.folderId)
    const tags = this.checkTags(options.tags ?? [])
    const order = 1 + Math.max(-1, ...this.inFolder(options.folderId ?? null).map((d) => d.order))
    const entry: DiagramEntry = { id, name: options.name ?? id, folderId: options.folderId ?? null, tags, order, excerpt: emptyExcerpt(), approvals: [] }
    this.diagrams.set(id, entry)
    if (options.model) this.update(id, options.model)
    return entry
  }

  private checkTags(tags: string[]): string[] {
    const unique = [...new Set(tags)]
    const unknown = unique.filter((tag) => !this.tags.has(tag))
    if (unknown.length) throw new CollectionError('unknownTags', { ids: unknown.join(', ') })
    return unique
  }

  entry(id: string): DiagramEntry {
    const entry = this.diagrams.get(id)
    if (!entry) throw new CollectionError('unknownDiagram', { id })
    return entry
  }

  /** Updates info and excerpt from the diagram model. */
  update(id: string, model: ProcessModel): DiagramEntry {
    const entry = this.entry(id)
    entry.info = model.info
    entry.excerpt = excerptFromModel(model)
    if (entry.name === entry.id && model.info?.title) entry.name = model.info.title
    return entry
  }

  rename(id: string, name: string): void {
    this.entry(id).name = name
  }

  remove(id: string): void {
    this.entry(id)
    this.diagrams.delete(id)
  }

  move(id: string, folderId: string | null, position?: number): void {
    const entry = this.entry(id)
    this.checkFolder(folderId)
    entry.folderId = folderId
    const siblings = this.inFolder(folderId).filter((d) => d.id !== id)
    siblings.splice(Math.max(0, Math.min(position ?? siblings.length, siblings.length)), 0, entry)
    siblings.forEach((item, index) => (item.order = index))
  }

  setOrder(folderId: string | null, ids: string[]): void {
    const current = new Set(this.inFolder(folderId).map((d) => d.id))
    if (ids.length !== current.size || !ids.every((id) => current.has(id))) throw new CollectionError('order')
    ids.forEach((id, index) => (this.entry(id).order = index))
  }

  setTags(id: string, tags: string[]): void {
    this.entry(id).tags = this.checkTags(tags)
  }

  // -- queries -----------------------------------------------------------------
  inFolder(folderId: string | null | undefined): DiagramEntry[] {
    return [...this.diagrams.values()].filter((d) => (d.folderId ?? null) === (folderId ?? null)).sort(byOrder)
  }

  subfolders(folderId: string | null | undefined): Folder[] {
    return [...this.folders.values()].filter((f) => (f.parentId ?? null) === (folderId ?? null)).sort(byOrder)
  }

  diagramsIn(folderId: string | null, recursive = true): DiagramEntry[] {
    const own = this.inFolder(folderId)
    return recursive ? [...own, ...this.subfolders(folderId).flatMap((f) => this.diagramsIn(f.id, true))] : own
  }

  withTag(tagId: string): DiagramEntry[] {
    return [...this.diagrams.values()].filter((d) => d.tags.includes(tagId))
  }

  tree(folderId: string | null = null): FolderNode {
    return {
      folder: folderId ? this.folder(folderId) : null,
      subfolders: this.subfolders(folderId).map((f) => this.tree(f.id)),
      diagrams: this.inFolder(folderId),
    }
  }

  /** Full-text search over name, title, subtitle, description and keywords. */
  search(term = '', options: { status?: string; tag?: string } = {}): DiagramEntry[] {
    const needle = term.trim().toLocaleLowerCase('de')
    return [...this.diagrams.values()]
      .filter((d) => !options.status || d.info?.status === options.status)
      .filter((d) => !options.tag || d.tags.includes(options.tag))
      .filter((d) => {
        const info = d.info
        const haystack = [d.name, info?.title, info?.subtitle, info?.description, ...(info?.keywords ?? [])].filter(Boolean).join(' ')
        return haystack.toLocaleLowerCase('de').includes(needle)
      })
      .sort((a, b) => a.name.localeCompare(b.name, 'de') || a.id.localeCompare(b.id))
  }

  /** `[diagramId, elementId]` of all elements carrying the domain key. */
  elementsForKey(kind: KeyKind, value: string): [string, string][] {
    if (!(KEY_KINDS as readonly string[]).includes(kind)) throw new CollectionError('unknownKind', { kind })
    const wanted = value.trim()
    return [...this.diagrams.values()]
      .sort((a, b) => a.id.localeCompare(b.id))
      .flatMap((d) => (d.excerpt.keys[kind]?.[wanted] ?? []).map((elementId): [string, string] => [d.id, elementId]))
  }

  // -- (de)serialisation --------------------------------------------------------
  toData(): CollectionData {
    const sorted = <T extends { id: string }>(values: Iterable<T>) => [...values].sort((a, b) => a.id.localeCompare(b.id))
    return {
      schema: COLLECTION_SCHEMA,
      id: this.id,
      name: this.name,
      folders: sorted(this.folders.values()).map((f) => ({ ...f })),
      tags: sorted(this.tags.values()).map((t) => ({ ...t })),
      diagrams: sorted(this.diagrams.values()).map((d) => structuredClone(d)),
    }
  }

  static fromData(data: CollectionData): DiagramCollection {
    if (data.schema !== COLLECTION_SCHEMA) throw new CollectionError('schema', { schema: String(data.schema) })
    const collection = new DiagramCollection(data.id, data.name)
    for (const folder of data.folders) collection.folders.set(folder.id, { ...folder })
    for (const tag of data.tags) collection.tags.set(tag.id, { ...tag })
    for (const entry of data.diagrams) collection.diagrams.set(entry.id, { ...structuredClone(entry), excerpt: { ...emptyExcerpt(), ...entry.excerpt } })
    for (const id of collection.folders.keys()) collection.ancestors(id)
    return collection
  }
}

File: packages/auditcore_bpmn/src/auditcore_bpmn/collection/collection.py
"""Diagrammsammlung: Ordner, Tags, Reihenfolge und Diagrammeinträge.

Jedes Diagramm liegt in genau einem Ordner (``folder_id``; ``None`` = oberste
Ebene) und trägt beliebig viele Tags. Aus dem XML hält die Sammlung nur einen
Auszug (:class:`DiagramExcerpt`). Abfragen, Übersichten, Verweise, Regeln und
Freigaben stehen in eigenen Modulen und sind hier als Methoden erreichbar.
"""

from __future__ import annotations

from collections.abc import Iterable
from datetime import date
from typing import TYPE_CHECKING

from ..errors import CollectionError
from ..model import BpmnDocument, as_document
from .excerpt import excerpt_from_document
from .model import Approval, DiagramEntry, DiagramReference, Folder, FolderNode, GroupOverview, Tag, check_id

if TYPE_CHECKING:
    from ..validation import ValidationIssue


def _next_position(positions: Iterable[int]) -> int:
    return 1 + max(positions, default=-1)


class DiagramCollection:
    """Veränderliche Sammlung; jede Änderung prüft die Integrität sofort."""

    def __init__(self, id: str = "sammlung", name: str = "Diagrammsammlung") -> None:  # noqa: A002
        self.id = id
        self.name = name
        self.folders: dict[str, Folder] = {}
        self.tags: dict[str, Tag] = {}
        self.diagrams: dict[str, DiagramEntry] = {}

    # -- Ordner ------------------------------------------------------------------
    def add_folder(
        self,
        id: str,
        name: str,
        parent_id: str | None = None,
        *,  # noqa: A002
        position: int | None = None,
        description: str | None = None,
    ) -> Folder:
        """Legt einen Ordner an (Position am Ende, falls nicht angegeben)."""
        check_id(id, "Ordner")
        if id in self.folders:
            raise CollectionError(f"Ordner „{id}“ existiert bereits.")
        if parent_id is not None and parent_id not in self.folders:
            raise CollectionError(f"Übergeordneter Ordner „{parent_id}“ ist unbekannt.")
        if position is None:
            position = _next_position(f.position for f in self.folders.values() if f.parent_id == parent_id)
        folder = Folder(id, name, parent_id, position, description)
        self.folders[id] = folder
        return folder

    def ancestors(self, folder_id: str | None) -> list[str]:
        """Kette bis zur obersten Ebene; wirft bei Zyklus oder unbekanntem Ordner."""
        chain: list[str] = []
        current = folder_id
        while current is not None:
            if current in chain:
                raise CollectionError(f"Ordner „{current}“ bildet einen Zyklus.")
            folder = self.folders.get(current)
            if folder is None:
                raise CollectionError(f"Ordner „{current}“ ist unbekannt.")
            chain.append(current)
            current = folder.parent_id
        return chain

    def move_folder(self, folder_id: str, parent_id: str | None) -> None:
        """Hängt einen Ordner um; Zyklen werden abgewiesen."""
        if folder_id not in self.folders:
            raise CollectionError(f"Ordner „{folder_id}“ ist unbekannt.")
        if parent_id is not None and folder_id in self.ancestors(parent_id):
            raise CollectionError("Ein Ordner kann nicht in sich selbst oder einen Unterordner verschoben werden.")
        self.folders[folder_id].parent_id = parent_id

    def remove_folder(self, folder_id: str) -> None:
        """Entfernt einen leeren Ordner."""
        occupied = any(f.parent_id == folder_id for f in self.folders.values()) or any(
            d.folder_id == folder_id for d in self.diagrams.values()
        )
        if occupied:
            raise CollectionError(f"Ordner „{folder_id}“ ist nicht leer.")
        self.folders.pop(folder_id, None)

    def add_tag(self, id: str, name: str, color: str | None = None) -> Tag:  # noqa: A002
        """Legt einen Tag an oder ersetzt ihn."""
        check_id(id, "Tag")
        tag = Tag(id, name, color)
        self.tags[id] = tag
        return tag

    # -- Diagramme ------------------------------------------------------------------
    def _check_tags(self, tags: Iterable[str]) -> list[str]:
        tag_ids = list(dict.fromkeys(tags))
        unknown = [t for t in tag_ids if t not in self.tags]
        if unknown:
            raise CollectionError(f"Unbekannte Tags: {unknown}")
        return tag_ids

    def add_diagram(
        self,
        id: str,
        xml: str | bytes | BpmnDocument | None = None,
        *,  # noqa: A002
        name: str | None = None,
        folder_id: str | None = None,
        tags: Iterable[str] = (),
        position: int | None = None,
    ) -> DiagramEntry:
        """Nimmt ein Diagramm auf; mit XML werden Infos und Auszug übernommen."""
        check_id(id, "Diagramm")
        if id in self.diagrams:
            raise CollectionError(f"Diagramm „{id}“ existiert bereits.")
        if folder_id is not None and folder_id not in self.folders:
            raise CollectionError(f"Ordner „{folder_id}“ ist unbekannt.")
        tag_ids = self._check_tags(tags)
        if position is None:
            position = _next_position(d.position for d in self.diagrams.values() if d.folder_id == folder_id)
        entry = DiagramEntry(id, name or id, folder_id, tag_ids, position)
        self.diagrams[id] = entry
        if xml is not None:
            self.refresh(id, xml)
        return entry

    def entry(self, diagram_id: str) -> DiagramEntry:
        """Eintrag eines Diagramms oder :class:`CollectionError`."""
        try:
            return self.diagrams[diagram_id]
        except KeyError as error:
            raise CollectionError(f"Diagramm „{diagram_id}“ ist unbekannt.") from error

    def refresh(self, diagram_id: str, xml: str | bytes | BpmnDocument) -> DiagramEntry:
        """Übernimmt Diagramm-Infos und Auszug aus dem XML."""
        entry = self.entry(diagram_id)
        document = as_document(xml)
        entry.info = document.diagram_info
        entry.excerpt = excerpt_from_document(document)
        if entry.name == entry.id and entry.info and entry.info.title:
            entry.name = entry.info.title
        return entry

    def remove_diagram(self, diagram_id: str) -> None:
        """Entfernt ein Diagramm aus der Sammlung."""
        self.entry(diagram_id)
        del self.diagrams[diagram_id]

    def move_diagram(self, diagram_id: str, folder_id: str | None, *, position: int | None = None) -> None:
        """Verschiebt ein Diagramm in einen Ordner (optional an eine Position) und nummeriert neu."""
        entry = self.entry(diagram_id)
        if folder_id is not None and folder_id not in self.folders:
            raise CollectionError(f"Ordner „{folder_id}“ ist unbekannt.")
        entry.folder_id = folder_id
        siblings = [d for d in self.in_folder(folder_id) if d.id != diagram_id]
        index = len(siblings) if position is None else max(0, min(position, len(siblings)))
        siblings.insert(index, entry)
        for number, item in enumerate(siblings):
            item.position = number

    def set_order(self, folder_id: str | None, diagram_ids: list[str]) -> None:
        """Setzt die Reihenfolge aller Diagramme eines Ordners."""
        current = {d.id for d in self.in_folder(folder_id)}
        if set(diagram_ids) != current or len(diagram_ids) != len(current):
            raise CollectionError("Die Reihenfolge muss genau die Diagramme des Ordners enthalten.")
        for number, diagram_id in enumerate(diagram_ids):
            self.diagrams[diagram_id].position = number

    def set_tags(self, diagram_id: str, tags: Iterable[str]) -> None:
        """Ersetzt die Tags eines Diagramms (nur bekannte Tags)."""
        self.entry(diagram_id).tags = self._check_tags(tags)

    # -- Einfache Abfragen ------------------------------------------------------------
    def in_folder(self, folder_id: str | None) -> list[DiagramEntry]:
        """Diagramme direkt in einem Ordner, sortiert nach Position und Name."""
        items = (d for d in self.diagrams.values() if d.folder_id == folder_id)
        return sorted(items, key=lambda d: (d.position, d.name.casefold(), d.id))

    def subfolders(self, folder_id: str | None) -> list[Folder]:
        """Unterordner eines Ordners, sortiert nach Position und Name."""
        items = (f for f in self.folders.values() if f.parent_id == folder_id)
        return sorted(items, key=lambda f: (f.position, f.name.casefold(), f.id))

    def diagrams_in(self, folder_id: str | None, *, recursive: bool = True) -> list[DiagramEntry]:
        """Diagramme eines Ordners, auf Wunsch einschließlich Unterordnern."""
        result = list(self.in_folder(folder_id))
        if recursive:
            for folder in self.subfolders(folder_id):
                result.extend(self.diagrams_in(folder.id))
        return result

    def with_tag(self, tag_id: str) -> list[DiagramEntry]:
        """Diagramme mit einem Tag."""
        return [d for d in self.diagrams.values() if tag_id in d.tags]

    def tree(self, folder_id: str | None = None) -> FolderNode:
        """Ordnerbaum ab ``folder_id`` (``None`` = oberste Ebene)."""
        return FolderNode(
            self.folders.get(folder_id) if folder_id else None,
            tuple(self.tree(f.id) for f in self.subfolders(folder_id)),
            tuple(self.in_folder(folder_id)),
        )

    # -- Delegationen ------------------------------------------------------------
    def search(self, term: str = "", *, status: str | None = None, tag: str | None = None) -> list[DiagramEntry]:
        """Volltextsuche, siehe :func:`auditcore_bpmn.collection.queries.search`."""
        from .queries import search

        return search(self, term, status=status, tag=tag)

    def elements_by_key(self, kind: str, value: str) -> list[tuple[str, str]]:
        """Elemente zu einem fachlichen Schlüssel, siehe :func:`queries.elements_by_key`."""
        from .queries import elements_by_key

        return elements_by_key(self, kind, value)

    def overview(
        self,
        folder_id: str | None = None,
        *,
        recursive: bool = True,
        tag_id: str | None = None,
        reference_date: date | None = None,
    ) -> GroupOverview:
        """Übersicht einer Gruppe, siehe :func:`queries.overview`."""
        from .queries import overview

        return overview(self, folder_id, recursive=recursive, tag_id=tag_id, reference_date=reference_date)

    def references(self) -> list[DiagramReference]:
        """Verweise zwischen Diagrammen, siehe :func:`queries.references`."""
        from .queries import references

        return references(self)

    def target_actual_pairs(self) -> list[tuple[str, str]]:
        """Paare aus Ist- und Soll-Diagramm."""
        from .queries import target_actual_pairs

        return target_actual_pairs(self)

    def validate(self) -> list[ValidationIssue]:
        """Regeln der Sammlung (``BPMN-K…``)."""
        from .checks import validate_collection

        return validate_collection(self)

    def approve(self, diagram_id: str, xml: str | bytes, version: str, **details: str | None) -> Approval:
        """Hält einen freigegebenen Stand fest, siehe :func:`approvals.approve`."""
        from .approvals import approve

        return approve(self, diagram_id, xml, version, **details)

    def check_approval(self, diagram_id: str, xml: str | bytes, version: str | None = None) -> ValidationIssue | None:
        """Prüft XML gegen einen freigegebenen Stand, siehe :func:`approvals.check_approval`."""
        from .approvals import check_approval

        return check_approval(self, diagram_id, xml, version)

    def to_dict(self) -> dict[str, object]:
        """JSON-Daten nach ``diagram-collection-1.schema.json``."""
        from .serialization import collection_to_dict

        return collection_to_dict(self)

    def to_json(self) -> str:
        """JSON-Text (UTF-8, eingerückt)."""
        from .serialization import collection_to_json

        return collection_to_json(self)

    @classmethod
    def from_dict(cls, data: dict[str, object]) -> DiagramCollection:
        """Sammlung aus JSON-Daten; prüft Schema, IDs und Ordnerzyklen."""
        from .serialization import collection_from_dict

        return collection_from_dict(data)

    @classmethod
    def from_json(cls, text: str) -> DiagramCollection:
        """Sammlung aus JSON-Text."""
        from .serialization import collection_from_json

        return collection_from_json(text)
